srec2fw.1
.\"
.\" This man page was written by Tormod Volden for Debian
.\" 
.TH SREC2FW "1"
.SH NAME
srec2fw \- convert from SREC format to linux kernel format
.SH SYNOPSIS
.B srec2fw
[\fIOPTION\fR]...
[\fIsrc.hex\fR] [\fIdst.fw\fR] 
.SH DESCRIPTION
Converts srec firmware files into binary representation for use by Linux kernel.
.PP
If the source file is `-' stdin will be read, and if destination file is `-'
the converted data will be written to stdout.
.TP
\fB\-w\fR
Source file uses wide records (16-bit length)
.SH "RETURN VALUE"
Returns 0 if no errors were encountered.
.SH EXAMPLES
.B srec2fw firmware.hex firmware.fw
.SH AUTHOR
Karl Relton <karllinuxtest.relton@ntlworld.com>,
David Woodhouse <dwmw2@infradead.org>,
Jan Harkes <jaharkes@cs.cmu.edu>



NEWS
2016-06-17  Kamil Prusko  <kamilprusko@gmail.com>
	Release version 0.12.1:

	* Fix for build error

2016-06-11  Kamil Prusko  <kamilprusko@gmail.com>
	Release version 0.12.0:

	* Support for more desktops
	* Ability to pause the timer
	* Ability add plugins
	* Code refactoring
	* Bug fixes
	* Dropped Telepathy and Skype integration
	* Dropped support for GNOME Shell 3.14 and older

2016-05-06  Kamil Prusko  <kamilprusko@gmail.com>
	Release version 0.11.3:

	* Ignore small mouse movements in screen notification
	* Fixed indicator not updating
	* Cleaned up some compilation warnings (thanks @aerostitch)

2016-03-24  Kamil Prusko  <kamilprusko@gmail.com>
	Release version 0.11.2:

	* Support for GNOME Shell 3.20

2015-10-24  Kamil Prusko  <kamilprusko@gmail.com>
	Release version 0.11.1:

	* Support for GNOME Shell 3.18
	* Turkish translation (thanks @ErkanMDR)

2015-06-13  Kamil Prusko  <kamilprusko@gmail.com>
	Release version 0.11.0:

	* Support for GNOME Shell 3.14 and 3.16
	* Ability to change indicator appearance
	* Skype and Epiphany integration
	* Updated French translation (thanks @neitsab)
	* Code refactoring
	* Basic unit tests
	* Bug fixes

2014-10-17  Kamil Prusko  <kamilprusko@gmail.com>
	Release version 0.10.3:

	* Support for GNOME Shell 3.10, 3.12 and 3.14
	* Updated Catalan translation (thanks @Ecron)
	* Code cleanups (thanks @aerostitch)
	* Bug fixes

2014-06-13  Kamil Prusko  <kamilprusko@gmail.com>

	Release version 0.10.2:

	* No need to restart gnome-shell to enable extension

2014-05-02  Kamil Prusko  <kamilprusko@gmail.com>

	Release version 0.10.1:

	* Support for GNOME Shell 3.10 and 3.12
	* Louder sounds
	* Fixed brining preferences dialog to focus
	* Fixed change of notications volume
	* Fixed reminders being showed up during pomodoro

2014-01-11  Kamil Prusko  <kamilprusko@gmail.com>

	Release version 0.10.0:

	* Support for GNOME Shell 3.10
	* New layout in preferences dialog
	* Migrate to gsteramer-1.0
	* Updated translations

2013-12-30  Kamil Prusko  <kamilprusko@gmail.com>

	Release version 0.9.1:

	* Support for GNOME Shell 3.8
	* Improved long pause scheduling
	* Deactivate screensaver to notify start of pomodoro
	* German translation (thanks @linuxrider)
	* Bug fixes

2013-11-17  Kamil Prusko  <kamilprusko@gmail.com>

	Release version 0.9.0:

	* Support for GNOME Shell 3.8
	* Added a preferences dialog
	* Improved timer accuracy
	* Bug fixes

2013-10-18  Kamil Prusko  <kamilprusko@gmail.com>

	Release version 0.8.1:

	* Support for GNOME Shell 3.10

2013-04-28  Kamil Prusko  <kamilprusko@gmail.com>

	Release version 0.8:

	* Support for GNOME Shell 3.8 (thanks @haaja)
	* Brazilian Portuguese translation (thanks @aleborba)
	* Minor bug fixes

2013-01-15  Kamil Prusko  <kamilprusko@gmail.com>

	Release version 0.7:

	* Support for GNOME Shell 3.4 and 3.6
	* Feature: Full screen notifications
	* Feature: Reminders
	* Chinese translation (thanks @mengzhuo)
	* Czech translation (thanks @veverjak)

2012-04-13  Arun Mahapatra  <pratikarun@gmail.com>

	Release version 0.6:

	* Support for GNOME Shell 3.4
	* Breaking change: Dropped support for older gnome-shell versions due to
	incompatible APIs
	* Feature: Support for "Away from desk" mode
	* Feature: Ability to change IM presence status based on pomodoro activity
	* New translation: Persian (thanks @arashm)
	* Fixed issues #38, #39, #41, #42, #45 and
	[more](https://github.com/codito/gnome-shell-pomodoro/issues?sort=created&direction=desc&state=closed&page=1)

2011-12-04  Arun  <pratikarun@gmail.com>

	Release version 0.5:

	* Bunch of cleanups, user interface awesomeness [Issue #37, Patch from @kamilprusko]
	* Config options are changed to more meaningful names [above patch]

2011-11-20  Arun  <pratikarun@gmail.com>

	Release version 0.4:

	* Sound notification at end of a pomodoro break [Issue #26, Patch from @kamilprusko]
	* System wide config file support [Patch from @mgrela]
	* Support to skip breaks in case of persistent message [Patch from @amanbh]
	* Some minor bug fixes, and keybinder3 requirement is now optional


srec2fw.c
/*
 * Parser/loader for SREC formatted data.
 *
 * Adapted from ihex2fw.c
 * 
 * Copyright © 2009 Karl Relton <karllinuxtest.relton@ntlworld.com>
 * Copyright © 2008 David Woodhouse <dwmw2@infradead.org>
 * Copyright © 2005 Jan Harkes <jaharkes@cs.cmu.edu>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 */

#include <stdint.h>
#include <arpa/inet.h>
#include <stdio.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#define _GNU_SOURCE
#include <getopt.h>

/* Define a special address for S7 start address records */
#define S3ADDR_START		(0xff400000UL)


struct srec_binrec {
	struct srec_binrec *next; /* not part of the real data structure */
        uint32_t addr;
        uint16_t len;
        uint8_t data[];
};

/**
 * nybble/hex are little helpers to parse hexadecimal numbers to a byte value
 **/
static uint8_t nybble(const uint8_t n)
{
       if      (n >= '0' && n <= '9') return n - '0';
       else if (n >= 'A' && n <= 'F') return n - ('A' - 10);
       else if (n >= 'a' && n <= 'f') return n - ('a' - 10);
       return 0;
}

static uint8_t hex(const uint8_t *data, uint8_t *crc)
{
       uint8_t val = (nybble(data[0]) << 4) | nybble(data[1]);
       *crc += val;
       return val;
}

static int process_srec(uint8_t *data, ssize_t size);
static void file_record(struct srec_binrec *record);
static int output_records(int outfd);

static int wide_records = 0;

int usage(void)
{
	fprintf(stderr, "srec2fw: Convert srec files into binary "
		"representation for use by Linux kernel\n");
	fprintf(stderr, "usage: srec2fw [<options>] <src.hex> <dst.fw>\n");
	fprintf(stderr, "       -w: wide records (16-bit length)\n");
	return 1;
}

int main(int argc, char **argv)
{
	int infd, outfd;
	struct stat st;
	uint8_t *data;
	int opt;

	while ((opt = getopt(argc, argv, "ws")) != -1) {
		switch (opt) {
		case 'w':
			wide_records = 1;
			break;
		default:
			return usage();
		}
	}

	if (optind + 2 != argc)
		return usage();

	if (!strcmp(argv[optind], "-"))
	    infd = 0;
	else
		infd = open(argv[optind], O_RDONLY);
	if (infd == -1) {
		fprintf(stderr, "Failed to open source file: %s",
			strerror(errno));
		return usage();
	}
	if (fstat(infd, &st)) {
		perror("stat");
		return 1;
	}
	data = mmap(NULL, st.st_size, PROT_READ, MAP_SHARED, infd, 0);
	if (data == MAP_FAILED) {
		perror("mmap");
		return 1;
	}

	if (!strcmp(argv[optind+1], "-"))
	    outfd = 1;
	else
		outfd = open(argv[optind+1], O_TRUNC|O_CREAT|O_WRONLY, 0644);
	if (outfd == -1) {
		fprintf(stderr, "Failed to open destination file: %s",
			strerror(errno));
		return usage();
	}
	if (process_srec(data, st.st_size))
		return 1;

	output_records(outfd);
	return 0;
}

static int process_srec(uint8_t *data, ssize_t size)
{
	struct srec_binrec *record, *record_new;
	uint32_t offset = 0;
	uint8_t type, crc, crcbyte = 0;
	int i, j;
	int line = 1;
	int len;

	i = 0;
next_record:
	/* search for the start of record character */
	while (i < size) {
		if (data[i] == '\n') line++;
		if (data[i++] == 'S') break;
	}

	if (i >= size) return 0;
    
	/* Minimum record length would be about 10 characters */
	if (i + 10 > size) {
		fprintf(stderr, "Can't find valid record at line %d\n", line);
		return -EINVAL;
	}

	crc = 1; /* CRC seeded with 1 for each line */
	type = nybble(data[i]); i++;
	len = hex(data + i, &crc); i += 2;
	if (wide_records) {
		len <<= 8;
		len += hex(data + i, &crc); i += 2;
	}
	len = len - 5;
	record = malloc((sizeof (*record) + len + 3) & ~3);
	if (!record) {
		fprintf(stderr, "out of memory for records\n");
		return -ENOMEM;
	}
	memset(record, 0, (sizeof(*record) + len + 3) & ~3);
	record->len = len;

	/* now check if we have enough data to read everything */
	if (i + 10 + (record->len * 2) > size) {
		fprintf(stderr, "Not enough data to read complete record at line %d\n",
			line);
		return -EINVAL;
	}

	record->addr  = hex(data + i, &crc) << 24; i += 2;
	record->addr |= hex(data + i, &crc) << 16; i += 2;
	record->addr |= hex(data + i, &crc) << 8; i += 2;
	record->addr |= hex(data + i, &crc); i += 2;

	for (j = 0; j < record->len; j++, i += 2)
		record->data[j] = hex(data + i, &crc);

	/* check CRC */
	crcbyte = hex(data + i, &crc); i += 2;
	if (crc != 0) {
		fprintf(stderr, "CRC failure at line %d: got 0x%X, expected 0x%X\n",
			line, crcbyte, (unsigned char)(crcbyte-crc));
		return -EINVAL;
	}

	/* Done reading the record */
	switch (type) {
	case 7:
		/* S7 Start Address Record - we will mutate this
		   into a new S3 record, with the address becoming the
		   data portion */
		record_new = malloc((sizeof (*record) + 4 + 3) & ~3);
		if (!record_new) {
			fprintf(stderr, "out of memory for records\n");
			return -ENOMEM;
		}
		memset(record_new, 0, (sizeof(*record) + 4 + 3) & ~3);
		record_new->len = 4;
		record_new->addr = S3ADDR_START;
		record->addr = htole32(record->addr);
		memcpy(record_new->data, &(record->addr), 4);
		file_record(record_new);
		goto next_record;

	case 3:
		/* S3 Address Record */
		file_record(record);
		goto next_record;

	default:
		fprintf(stderr, "Unknown record (type %02X)\n", type);
	}

	return -EINVAL;
}

static struct srec_binrec *records;

static void file_record(struct srec_binrec *record)
{
	struct srec_binrec **p = &records;

	while ((*p) && ((*p)->addr < record->addr))
		p = &((*p)->next);

	record->next = *p;
	*p = record;
}

static int output_records(int outfd)
{
	unsigned char zeroes[6] = {0, 0, 0, 0, 0, 0};
	struct srec_binrec *p = records;

	while (p) {
		uint16_t writelen = (p->len + 9) & ~3;

		p->addr = htonl(p->addr);
		p->len = htons(p->len);
		write(outfd, &p->addr, writelen);
		p = p->next;
	}
	/* EOF record is zero length, since we don't bother to represent
	   the type field in the binary version */
	write(outfd, zeroes, 6);
	return 0;
}
